### ISSUE TITLE
refetchOnFocus removing cached values despite 'keepUnusedDataFor' being set.

### ISSUE THREAD

**Issue by mattoni** [reduxjs/redux-toolkit#3117]
I have a situation that seems unexpected to me. I have a query with 'keepUnusedDataFor' set to 30 minutes (the length of something expiring in the API). I also have 'refetchOnFocus: true' set for my whole API, but disabled for this specific query.

If I keep focus on the window, the token lasts in cache like I want despite no subscriptions to it. However, if I lose focus and later (within the 30 minutes) regain focus, my value is cleared out of cache. No new fetch is done, as expected, but losing the value on focus is not what I intend.

Is this a bug, or is there something I'm not configuring correctly?

The query: 
```
    const [getSshCreds, sshCreds] = useLazyGetSshConnectionQuery({
        refetchOnFocus: false,
        refetchOnReconnect: false,
    });
```

The settings:
![image](https://user-images.githubusercontent.com/5110855/214168095-93b00740-2f49-41f8-a29d-568eb8514861.png)

The api service:
![image](https://user-images.githubusercontent.com/5110855/214167806-105554a5-851c-4a85-ac6a-f96773ccaa1f.png)

**Comment by markerikson** [comment]
Hmm. I'm kind of confused with the setup and description.

Could you put together a CodeSandbox or repo that shows this happening? Alternately, could you provide a Replay ( https://replay.io/record-bugs ) of this occurring on your machine?  (you'd probably want to lower the timer to, say, 10-15 seconds instead of 30 minutes :)  )

**Comment by mattoni** [comment]
Here's a quick and dirty repo.

https://codesandbox.io/s/elastic-kapitsa-1n1lxb?file=/src/TestComponent.tsx

 I found with codesandbox you'll want to expand it to a full window (https://1n1lxb.csb.app/)


To verify, open redux toolkit and check that the subscription was created. I then unmount this component 2 seconds after launch, so the data is 'kept' in the store, but there are no subscriptions. This is fine, but then unfocus the window, then refocus. Upon refocus, the data is lost.

**Comment by phryneas** [comment]
It is working as intended.

"Refetch" here essentially gives you the guarantee of "all data in cache is up to date". Now, for unsubscribed data, we have three options: 

* refetch all unsubscribed cache entries. This could be hundreds of unnneccessary requests.
* keep them as they are. That would destroy the guarantee you expect from a "refetch" - your data is not up to date
* remove them from cache - if they component remounts, they can be fetched from the server again. But if the component does not remount, there is no wasted request to the server.

As you see, of these three, the only viable option is the third one.

**Comment by mattoni** [comment]
hmm. is there no way to hint 'keep this data around, i specifically am asking to keep it and not update it'? i suppose i could funnel it to another key in the store but that seems like a lot of extra work to do something that RTKQ already takes care of.

In the API i'm hitting, a 'get' request to this endpoint generates new SSH credentials. The credentials automatically expire after 30 minutes, but I don't want to keep generating new ones every time I need them. I could use a skip token, but i'd need to keep the expiration somewhere else, that is then required to be manually managed etc. If there's no other way I can do it that way, but if there's a way in RTKQ to ensure I keep this data until its expired, I'd prefer that.

**Comment by markerikson** [comment]
@mattoni : that's explicitly the purpose of the `keepUnusedDataFor` field - how long you want to keep it around when no component is actually _subscribed_ to that data any more.

**Comment by mattoni** [comment]
Yes - that is what I've done in the above example (set for 30 minutes). However, refocusing removes it despite this being set, which is why I was confused.

**Comment by markerikson** [comment]
It looks like the "remove based on focus change" behavior is "intentional", at least in the sense of "that's what the current code is _written_ to do":

https://github.com/reduxjs/redux-toolkit/blob/d2d0f708abe3c51ae68bd5a96d83826f8897c355/packages/toolkit/src/query/core/buildMiddleware/windowEventHandling.ts#L37-L62

**Comment by mattoni** [comment]
@phryneas the `refetchOnFocus` ignoring `keepUnusedDataFor` value is why it feels unexpected. I realize my example was flawed in that it wasn't exactly how I was using it (using a lazy query and overriding refetch) so I updated it to more accurately represent my use case.

As for the mutation part, I agree that would work as well. I just expected the  `refetchOnFocus`  to respect the `keepUnusedDataFor` , since in a sense, it's sort of a subscription (or an override for one).

**Comment by phryneas** [comment]
> As for the mutation part, I agree that would work as well. I just expected the refetchOnFocus to respect the keepUnusedDataFor , since in a sense, it's sort of a subscription (or an override for one).

It's not a subscription, but the number of seconds something will be kept if there is no subscription - unless something (like a validation or refetch) happens   
Maybe that gets a bit more clear by looking at the default value: not 0, but 60 seconds. If it would force things to stay in cache, as explained above you would either massively overfetch or keep stale values in cache.

**Comment by mattoni** [comment]
I see, that makes sense thank you.

**Comment by mattoni** [comment]
Treating it as a mutation +fixedCacheKey worked like a charm, thanks!

**Comment by kasparkallas** [comment]
The fact that `refetchOnFocus` removes unused data came as a surprise for me too. I guess it does make sense from correctness perspective if you think about it. It's just not the magical property to solve all of my problems that I thought it was.

This was my hopeful expectation:
1. Refetches subscribed data on focus
2. Doesn't remove unused data
3. Refetches unused data on resubscription but uses the unused data to bootstrap the subscription

**Comment by markerikson** [comment]
Yeah, I think there's some discussion worth having here (although given the number of things we're juggling atm this may not be an immediate priority).

**Comment by kasparkallas** [comment]
Thoughts on adding `refetchOnFocus_keepUnusedData` & `refetchOnReconnect_keepUnusedData` options if I were to do a PR?

Or would it be better to make `refetchOnFocus` also specifiable with an object:
```ts
{
   refetchOnFocus: {
     keepUnusedData: true
   }
}
``` 

While I'm bringing this up, anything that has got something to do with refetching or rehydrating then I would kind of want to see the `ifOlderThan` option too. Because if the user is rapidly changing tabs back-and-forth or fiddling with a browser extension, the `refetchOnFocus` could fire unexpectedly many refetches in short period of time. So it could possibly be:
```ts
{
   refetchOnFocus: {
     ifOlderThan: 15,
     keepUnusedData: true
   }
}
```
